This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [65]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case


Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [66]:
# First, import the relevant modules

import requests
import pandas as pd
import collections
import json
from pandas import DataFrame

In [67]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned


data = r.json()

        

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [68]:

for key, value in data.items():
#     print(key)
    inner_dict  = value
    for key2, value2 in inner_dict.items():
        pass
#         print('key:' + key2)
#         print('value:')
#         print(value2)

selected_data = data['dataset']['data']
# print(selected_data)
selected_data = DataFrame (selected_data,columns=data['dataset']['column_names'])
print(selected_data.head())

         Date   Open   High    Low  Close  Volume  Adjustment Factor  \
0  2018-11-30  17.54  17.54  17.54  17.54     0.0                NaN   
1  2018-11-29  17.04  17.04  17.04  17.04     0.0                NaN   
2  2018-11-28  17.38  17.38  17.38  17.38     0.0                NaN   
3  2018-11-27  17.60  17.60  17.60  17.60     0.0                NaN   
4  2018-11-26  17.54  17.54  17.54  17.54     0.0                NaN   

   Adjustment Type  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  


In [69]:
selected_year = selected_data[selected_data['Date'].str.contains("2017")]
selected_year["change"] = selected_year["High"] - selected_year["Low"]


print(selected_year.describe())

             Open        High         Low       Close       Volume  \
count  260.000000  260.000000  260.000000  260.000000   260.000000   
mean   106.124842  107.308138  105.854323  106.918096   131.357692   
std      6.613434    6.753165    6.461637    6.720764   299.410629   
min     91.000000   91.000000   91.000000   91.000000     0.000000   
25%    101.691750  102.650000  101.506750  102.487500     0.000000   
50%    105.563500  107.000000  105.492000  106.750000    50.000000   
75%    111.015750  112.502500  110.985750  112.017750   151.250000   
max    126.000000  126.788000  123.550000  126.788000  3700.000000   

       Adjustment Factor  Adjustment Type      change  
count           2.000000         2.000000  260.000000  
mean            0.981928        10.000000    1.453815  
std             0.025557         9.899495    1.654781  
min             0.963857         3.000000    0.000000  
25%             0.972893         6.500000    0.000000  
50%             0.981928        1

<ipython-input-69-d9fb9e46243d>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  selected_year["change"] = selected_year["High"] - selected_year["Low"]


In [70]:
selected_year_shift = pd.concat([selected_year, selected_year.shift(2)], axis=1)


selected_year_shift.columns = selected_year.columns.tolist()+(selected_year.columns+'_shiftback2').tolist()


selected_year_shift.change = selected_year_shift.Close - selected_year_shift.Close_shiftback2
selected_year_shift.head()

print(selected_year_shift.describe())

             Open        High         Low       Close       Volume  \
count  260.000000  260.000000  260.000000  260.000000   260.000000   
mean   106.124842  107.308138  105.854323  106.918096   131.357692   
std      6.613434    6.753165    6.461637    6.720764   299.410629   
min     91.000000   91.000000   91.000000   91.000000     0.000000   
25%    101.691750  102.650000  101.506750  102.487500     0.000000   
50%    105.563500  107.000000  105.492000  106.750000    50.000000   
75%    111.015750  112.502500  110.985750  112.017750   151.250000   
max    126.000000  126.788000  123.550000  126.788000  3700.000000   

       Adjustment Factor  Adjustment Type      change  Open_shiftback2  \
count           2.000000         2.000000  258.000000       258.000000   
mean            0.981928        10.000000    0.084043       106.065151   
std             0.025557         9.899495    2.485445         6.603604   
min             0.963857         3.000000   -8.938000        91.000000   

Here are the answers: Note this is a different stock because the stock price of Zeiss was not accessible. 

Calculate what the highest and lowest opening prices were for the stock in this period.
highest opening in 2017 is 126, lowest is 91

What was the largest change in any one day (based on High and Low price)?
largest change was 11.36

What was the largest change between any two days (based on Closing Price)?
largest change between two days was 11.36. 

What was the average daily trading volume during this year?
avg daily volume is 131.4

(Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)
median is 50# 